patches/fips.patch

```diff
@@ -127,32 +182,102 @@ index b7c26f91e5..0e6680fd04 100644
+@@ -9,6 +9,7 @@ import (
  	"crypto"
  	"crypto/cipher"
  	"crypto/ecdh"
--	"crypto/internal/boring"
 +	boring "crypto/internal/backend"
  	"crypto/rand"
  	"crypto/sha256"
  	"encoding/hex"
-@@ -288,8 +288,8 @@ func TestNewPrivateKey(t *testing.T) {
+@@ -37,6 +38,9 @@ var _ interface {
+ 
+ func TestECDH(t *testing.T) {
+ 	testAllCurves(t, func(t *testing.T, curve ecdh.Curve) {
++		if boring.Enabled() && curve == ecdh.X25519() {
++			t.Skip("curve unsupported in FIPS mode, negative test included elsewhere")
++		}
+ 		aliceKey, err := curve.GenerateKey(rand.Reader)
+ 		if err != nil {
+ 			t.Fatal(err)
+@@ -96,18 +100,29 @@ func (r *countingReader) Read(p []byte) (int, error) {
+ 
+ func TestGenerateKey(t *testing.T) {
+ 	testAllCurves(t, func(t *testing.T, curve ecdh.Curve) {
+-		r := &countingReader{r: rand.Reader}
++		rand := rand.Reader
++		var r io.Reader = &countingReader{r: rand}
++		if boring.Enabled() {
++			rand = boring.RandReader
++			r = rand
++		}
+ 		k, err := curve.GenerateKey(r)
+ 		if err != nil {
++			if boring.Enabled() && curve == ecdh.X25519() && strings.Contains(err.Error(), "crypto/ecdh: use of X25519 is not allowed in FIPS 140-only mode") {
++				t.Skip("expected error in FIPS mode")
++			}
+ 			t.Fatal(err)
+ 		}
++		if boring.Enabled() {
++			return // the rest of the assertion does not apply in boring mode
++		}
+ 
+ 		// GenerateKey does rejection sampling. If the masking works correctly,
+ 		// the probability of a rejection is 1-ord(G)/2^ceil(log2(ord(G))),
+ 		// which for all curves is small enough (at most 2^-32, for P-256) that
+ 		// a bit flip is more likely to make this test fail than bad luck.
+ 		// Account for the extra MaybeReadByte byte, too.
+-		if got, expected := r.n, len(k.Bytes())+1; got > expected {
++		if got, expected := r.(*countingReader).n, len(k.Bytes())+1; got > expected {
+ 			t.Errorf("expected GenerateKey to consume at most %v bytes, got %v", expected, got)
+ 		}
+ 	})
+@@ -160,6 +175,9 @@ func TestVectors(t *testing.T) {
+ 		v := vectors[curve]
+ 		key, err := curve.NewPrivateKey(hexDecode(t, v.PrivateKey))
+ 		if err != nil {
++			if boring.Enabled() && curve == ecdh.X25519() && strings.Contains(err.Error(), "crypto/ecdh: use of X25519 is not allowed in FIPS 140-only mode") {
++				t.Skip("error expected in FIPS mode")
++			}
+ 			t.Fatal(err)
+ 		}
+ 		if !bytes.Equal(key.PublicKey().Bytes(), hexDecode(t, v.PublicKey)) {
+@@ -208,6 +226,9 @@ func TestX25519Failure(t *testing.T) {
+ }
+ 
+ func testX25519Failure(t *testing.T, private, public []byte) {
++	if boring.Enabled() {
++		t.Skip("X25519 not supported in FIPS mode")
++	}
+ 	priv, err := ecdh.X25519().NewPrivateKey(private)
+ 	if err != nil {
+ 		t.Fatal(err)
+@@ -280,15 +301,26 @@ var invalidPrivateKeys = map[ecdh.Curve][]string{
+ }
+ 
+ func TestNewPrivateKey(t *testing.T) {
++	// Skip on RHEL 8
++	if b, err := os.ReadFile("/etc/os-release"); err == nil {
++		if strings.Contains(string(b), "Red Hat Enterprise Linux 8") {
```

Comment:
Can't this be checked against openssl version (through `openssl.VersionText()`)?

Reply:
Yes, if this approach proves problematic we can always do that.